Index: security/coreconf/Linux.mk
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/Linux.mk,v
retrieving revision 1.18
diff -d -u -p -r1.18 Linux.mk
--- security/coreconf/Linux.mk	1 Feb 2005 23:39:18 -0000	1.18
+++ security/coreconf/Linux.mk	20 Jun 2005 07:32:29 -0000
@@ -56,6 +56,10 @@ ifeq ($(OS_TEST),m68k)
 	OS_REL_CFLAGS	= -DLINUX1_2 -D_XOPEN_SOURCE
 	CPU_ARCH	= m68k
 else		
+ifeq ($(OS_TEST),ppc64)
+	OS_REL_CFLAGS	= -DLINUX1_2 -D_XOPEN_SOURCE
+	CPU_ARCH	= ppc64
+else
 ifeq ($(OS_TEST),ppc)
 	OS_REL_CFLAGS	= -DLINUX1_2 -D_XOPEN_SOURCE
 	CPU_ARCH	= ppc
@@ -119,6 +123,7 @@ endif
 endif
 endif
 endif
+endif
 
 
 LIBC_TAG		= _glibc
