--- mozilla/config/system-headers.visibility	2005-06-17 17:03:42.000000000 -0400
+++ mozilla/config/system-headers	2006-10-26 22:40:49.000000000 -0400
@@ -66,6 +66,8 @@
 Button.h
 byteswap.h
 cairo.h
+cairo-xlib.h
+cairo-xlib-xrender.h
 callconv.h
 Carbon/Carbon.h
 CarbonEvents.h
@@ -176,7 +178,7 @@
 freetype/freetype.h
 freetype/ftcache.h
 freetype/ftglyph.h
-freetype/ftoutlin.h
+freetype/ftoutln.h
 freetype/ttnameid.h
 freetype/tttables.h
 fribidi/fribidi.h
@@ -192,6 +194,8 @@
 gdk/gdkregion.h
 gdk/gdkwindow.h
 gdk/gdkx.h
+gdk/gdkpango.h
+gdk/gdkalias.h
 gdk-pixbuf/gdk-pixbuf.h
 Gestalt.h
 getopt.h
@@ -212,7 +216,11 @@
 gtk/gtkcontainer.h
 gtk/gtkdialog.h
 gtk/gtkentry.h
+gtk/gtkfilechooser.h
+gtk/gtkfixed.h
 gtk/gtk.h
+gtk/gtkiconfactory.h
+gtk/gtkimage.h
 gtk/gtkimmulticontext.h
 gtk/gtkinvisible.h
 gtk/gtkmain.h
@@ -496,10 +504,12 @@
 photon/PtWebClient.h
 photon/PxImage.h
 pk11func.h
+pk11pub.h
 pkcs11t.h
 PLStringFuncs.h
 PMApplication.h
 pmddim.h
+png.h
 poll.h
 Polygon.h
 portable.h
@@ -874,6 +884,7 @@
 X11/Xutil.h
 xpt_struct.h
 xpt_xdr.h
+zlib.h
 zmouse.h
 sslt.h
 smime.h
--- mozilla/modules/oji/src/nsJVMManager.cpp.visibility	2005-08-22 10:09:59.000000000 -0400
+++ mozilla/modules/oji/src/nsJVMManager.cpp	2006-10-26 22:38:48.000000000 -0400
@@ -845,7 +845,7 @@
     return fStatus;
 }
 
-extern "C" nsresult JSJ_RegisterLiveConnectFactory(void);
+extern "C" NS_VISIBILITY_DEFAULT nsresult JSJ_RegisterLiveConnectFactory(void);
 
 PRBool
 nsJVMManager::MaybeStartupLiveConnect(void)
