Index: mozilla/browser/components/nsBrowserContentHandler.js
===================================================================
RCS file: /cvsroot/mozilla/browser/components/nsBrowserContentHandler.js,v
retrieving revision 1.12.2.23
diff -d -u -p -r1.12.2.23 nsBrowserContentHandler.js
--- mozilla/browser/components/nsBrowserContentHandler.js	14 Sep 2007 19:59:42 -0000	1.12.2.23
+++ mozilla/browser/components/nsBrowserContentHandler.js	13 Dec 2007 13:20:44 -0000
@@ -542,18 +542,14 @@ var nsBrowserContentHandler = {
   },
 
   get startPage() {
+    var formatter = Components.classes["@mozilla.org/toolkit/URLFormatterService;1"]
+                              .getService(Components.interfaces.nsIURLFormatter);
+ 
     var prefb = Components.classes["@mozilla.org/preferences-service;1"]
                           .getService(nsIPrefBranch);
 
-    var uri = prefb.getComplexValue("browser.startup.homepage",
-                                    nsIPrefLocalizedString).data;
+    var uri = formatter.formatURLPref("browser.startup.homepage");
 
-    if (!uri) {
-      prefb.clearUserPref("browser.startup.homepage");
-      uri = prefb.getComplexValue("browser.startup.homepage",
-                                  nsIPrefLocalizedString).data;
-    }
-                                
     var count;
     try {
       count = prefb.getIntPref("browser.startup.homepage.count");
@@ -564,8 +560,7 @@ var nsBrowserContentHandler = {
 
     for (var i = 1; i < count; ++i) {
       try {
-        var page = prefb.getComplexValue("browser.startup.homepage." + i,
-                                         nsIPrefLocalizedString).data;
+        var page = formatter.formatURLPref("browser.startup.homepage." + i);
         uri += "\n" + page;
       }
       catch (e) {
