diff -up firefox-33.0/mozilla-release/toolkit/library/libxul.mk.old firefox-33.0/mozilla-release/toolkit/library/libxul.mk
--- firefox-33.0/mozilla-release/toolkit/library/libxul.mk.old	2014-11-03 13:37:20.206772190 +0100
+++ firefox-33.0/mozilla-release/toolkit/library/libxul.mk	2014-11-03 13:36:56.966691951 +0100
@@ -168,7 +168,7 @@ endif
 
 ifdef MOZ_WIDGET_GTK
 ifdef MOZ_ENABLE_GTK3
-EXTRA_DSO_LDOPTS += $(filter-out -lgtk-3 -lgdk-3,$(TK_LIBS)) -lmozgtk_stub
+EXTRA_DSO_LDOPTS += $(filter-out -lgtk-3 -lgdk-3,$(TK_LIBS)) -L$(LIBXUL_DIST)/lib -lmozgtk_stub
 else
 EXTRA_DSO_LDOPTS += $(TK_LIBS)
 endif
