Index: browser/locales/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/browser/locales/Makefile.in,v
retrieving revision 1.1.2.24
diff -d -u -p -r1.1.2.24 Makefile.in
--- browser/locales/Makefile.in	22 Oct 2004 16:47:44 -0000	1.1.2.24
+++ browser/locales/Makefile.in	24 May 2005 17:37:59 -0000
@@ -97,6 +97,7 @@ install:: $(addprefix $(AB_CD)/profile/,
 install:: $(addprefix $(AB_CD)/profile/chrome/,$(PROFILE_CHROME))
 	$(SYSINSTALL) $(IFLAGS1) $^ $(DESTDIR)$(mozappdir)/defaults/profile/chrome
 
+ifeq ($(AB_CD),en-US)
 SEARCH_PLUGINS = $(shell cat $(srcdir)/$(AB_CD)/searchplugins/list.txt)
 SEARCH_PLUGIN_PREFIX = $(addprefix $(srcdir)/$(AB_CD)/searchplugins/,$(SEARCH_PLUGINS))
 
@@ -105,6 +106,7 @@ libs realchrome:: $(addsuffix .src,$(SEA
 
 install:: $(addsuffix .src,$(SEARCH_PLUGIN_PREFIX)) $(foreach base,$(SEARCH_PLUGIN_PREFIX),$(wildcard $(base).gif)$(wildcard $(base).png))
 	$(SYSINSTALL) $(IFLAGS1) $^ $(DESTDIR)$(mozappdir)/searchplugins
+endif
 
 # Let's imagine we wanted to create a language JARfile without registering it
 # the following targets only work with JARed chrome: flat chrome is not supported
