diff -up firefox-42.0/firefox-42.0/configure.in.old firefox-42.0/firefox-42.0/configure.in
--- firefox-42.0/firefox-42.0/configure.in.old	2015-10-27 01:10:02.000000000 +0100
+++ firefox-42.0/firefox-42.0/configure.in	2015-10-29 11:56:49.779827023 +0100
@@ -53,7 +53,7 @@ dnl ====================================
 MOZJPEG=62
 MOZPNG=10617
 NSPR_VERSION=4
-NSPR_MINVER=4.10.10
+NSPR_MINVER=4.10.8
 NSS_VERSION=3
 
 dnl Set the minimum version of toolkit libs used by mozilla
