Index: browser/locales/en-US/chrome/browser-region/region.properties
===================================================================
RCS file: /cvsroot/mozilla/browser/locales/en-US/chrome/browser-region/region.properties,v
retrieving revision 1.1.2.3
diff -d -u -p -0 -r1.1.2.3 region.properties
--- browser/locales/en-US/chrome/browser-region/region.properties	29 Aug 2004 07:29:59 -0000	1.1.2.3
+++ browser/locales/en-US/chrome/browser-region/region.properties	25 Sep 2004 03:07:18 -0000
@@ -3 +3 @@ general.useragent.contentlocale=US
-homePageDefault=http://www.mozilla.org/products/firefox/start/
+homePageDefault=file:///usr/share/doc/HTML/index.html
@@ -10 +10 @@ promoteURL=http://www.spreadfirefox.com/
-browser.startup.homepage=http://start.mozilla.org/firefox/
+browser.startup.homepage=file:///usr/share/doc/HTML/index.html
@@ -15 +15 @@ browser.search.defaulturl=http://www.goo
-startup.homepage_override_url=http://www.mozilla.org/products/firefox/start/
+startup.homepage_override_url=file:///usr/share/doc/HTML/index.html
Index: browser/app/profile/firefox.js
===================================================================
RCS file: /cvsroot/mozilla/browser/app/profile/firefox.js,v
retrieving revision 1.7.4.38
diff -d -u -p -r1.7.4.38 firefox.js
--- browser/app/profile/firefox.js	26 Oct 2004 22:44:47 -0000	1.7.4.38
+++ browser/app/profile/firefox.js	31 Oct 2004 08:28:37 -0000
@@ -141,10 +141,10 @@ pref("browser.shell.checkDefaultBrowser"
 // 0 = blank, 1 = home (browser.startup.homepage), 2 = last
 // XXXBlake Remove this stupid pref
 pref("browser.startup.page",                1);
-pref("browser.startup.homepage",	          "resource:/browserconfig.properties");
+pref("browser.startup.homepage",	          "chrome://browser-region/locale/region.properties");
 
 // These values are deliberately non-localizable for official builds.
-pref("browser.startup.homepage_reset",      "resource:/browserconfig.properties");
+pref("browser.startup.homepage_reset",      "chrome://browser-region/locale/region.properties");
 pref("browser.update.resetHomepage",        false);
 // "browser.startup.homepage_override" was for 4.x
 pref("browser.startup.homepage_override.1", false);
