Index: mozilla/modules/libpref/src/init/all.js
===================================================================
RCS file: /cvsroot/mozilla/modules/libpref/src/init/all.js,v
retrieving revision 3.515.2.3.2.23
diff -d -u -p -0 -r3.515.2.3.2.23 all.js
--- mozilla/modules/libpref/src/init/all.js	4 Nov 2004 23:39:57 -0000	3.515.2.3.2.23
+++ mozilla/modules/libpref/src/init/all.js	4 Dec 2004 07:00:57 -0000
@@ -70 +70 @@ pref("browser.display.use_document_color
-pref("browser.display.use_system_colors",   false);
+pref("browser.display.use_system_colors",   true);
