Index: browser/themes/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/browser/themes/Makefile.in,v
retrieving revision 1.1.2.2
diff -d -u -p -r1.1.2.2 Makefile.in
--- browser/themes/Makefile.in	17 Jun 2004 06:18:32 -0000	1.1.2.2
+++ browser/themes/Makefile.in	4 Nov 2004 17:46:36 -0000
@@ -45,14 +45,19 @@ include $(DEPTH)/config/autoconf.mk
 #
 # Theme Selection
 #
-# Windows, GNOME/Linux    Winstripe
+# Windows                 Winstripe
 # MacOS X                 Pinstripe
+# GNOME/Linux             Gnomestripe
 #
 
 ifneq (,$(filter mac cocoa,$(MOZ_WIDGET_TOOLKIT)))
 DIRS = pinstripe
 else
+ifneq (,$(filter windows,$(MOZ_WIDGET_TOOLKIT)))
 DIRS = winstripe
+else
+DIRS = gnomestripe
+endif
 endif
 
 include $(topsrcdir)/config/rules.mk
