? nsprpub/config/system_wrappers
Index: nsprpub/config/nspr-config.in
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/config/nspr-config.in,v
retrieving revision 1.2.2.2
diff -d -u -p -r1.2.2.2 nspr-config.in
--- nsprpub/config/nspr-config.in	28 Feb 2003 01:52:53 -0000	1.2.2.2
+++ nsprpub/config/nspr-config.in	4 Apr 2005 21:33:29 -0000
@@ -3,6 +3,7 @@
 prefix=@prefix@
 exec_prefix=@exec_prefix@
 exec_prefix_set=no
+libdir=@libdir@
 
 major_version=@MOD_MAJOR_VERSION@
 minor_version=@MOD_MINOR_VERSION@
@@ -95,7 +96,7 @@ if test "$echo_cflags" = "yes"; then
 fi
 
 if test "$echo_libs" = "yes"; then
-      libdirs=-L${exec_prefix}/lib
+      libdirs=-L${libdir}
       if test -n "$lib_plds"; then
 	libdirs="$libdirs -lplds${major_version}"
       fi
