diff -up firefox-33.0/mozilla-release/media/webrtc/trunk/webrtc/build/common.gypi.old firefox-33.0/mozilla-release/media/webrtc/trunk/webrtc/build/common.gypi
--- firefox-33.0/mozilla-release/media/webrtc/trunk/webrtc/build/common.gypi.old	2014-10-11 11:06:40.000000000 +0200
+++ firefox-33.0/mozilla-release/media/webrtc/trunk/webrtc/build/common.gypi	2014-10-15 08:43:34.746223846 +0200
@@ -235,19 +235,6 @@
         'defines': [
           'WEBRTC_ARCH_ARM',
         ],
-        'conditions': [
-          ['arm_version==7', {
-            'defines': ['WEBRTC_ARCH_ARM_V7',
-                        'WEBRTC_BUILD_NEON_LIBS'],
-            'conditions': [
-              ['arm_neon==1', {
-                'defines': ['WEBRTC_ARCH_ARM_NEON',],
-              }, {
-                'defines': ['WEBRTC_DETECT_ARM_NEON',],
-              }],
-            ],
-          }],
-        ],
       }],
       ['os_bsd==1', {
         'defines': [
