diff -up firefox-29.0/mozilla-release/gfx/ycbcr/chromium_types.h.chromium-types firefox-29.0/mozilla-release/gfx/ycbcr/chromium_types.h
--- firefox-29.0/mozilla-release/gfx/ycbcr/chromium_types.h.chromium-types	2014-04-22 07:27:47.000000000 +0200
+++ firefox-29.0/mozilla-release/gfx/ycbcr/chromium_types.h	2014-04-22 15:46:24.288677788 +0200
@@ -34,7 +34,11 @@ typedef uint32_t uint32;
 #define ARCH_CPU_ARM_FAMILY 1
 #define ARCH_CPU_ARMEL 1
 #define ARCH_CPU_32_BITS 1
-#elif defined(__ppc__) || defined(__powerpc) || defined(__PPC__)
+#elif defined(__powerpc64__)
+#define ARCH_CPU_PPC_FAMILY 1
+#define ARCH_CPU_PPC 1
+#define ARCH_CPU_64_BITS 1
+#elif defined(__ppc__) || defined(__powerpc__) || defined(__PPC__)
 #define ARCH_CPU_PPC_FAMILY 1
 #define ARCH_CPU_PPC 1
 #define ARCH_CPU_32_BITS 1
@@ -50,6 +54,14 @@ typedef uint32_t uint32;
 #define ARCH_CPU_AARCH64_FAMILY 1
 #define ARCH_CPU_AARCH64 1
 #define ARCH_CPU_64_BITS 1
+#elif defined(__s390x__)
+#define ARCH_CPU_S390_FAMILY 1
+#define ARCH_CPU_S390 1
+#define ARCH_CPU_64_BITS 1
+#elif defined(__s390__)
+#define ARCH_CPU_S390_FAMILY 1
+#define ARCH_CPU_S390 1
+#define ARCH_CPU_32_BITS 1
 #else
 #warning Please add support for your architecture in chromium_types.h
 #endif
